
# 程序终止

参考：[Program Termination](https://docs.microsoft.com/en-us/cpp/cpp/program-termination?view=vs-2019)

`c++`程序终止方式共有`3`种

1. 调用`exit`
2. 调用`abort`
3. 在`main`函数调用`return`表达式

## exit

参考：

[exit Function](https://docs.microsoft.com/en-us/cpp/cpp/exit-function?view=vs-2019)

[exit](http://www.cplusplus.com/reference/cstdlib/exit/?kw=exit)

声明在头文件`stdlib.h`，其参数值作为程序返回代码或退出代码。按照惯例，返回`0`表示成功退出

>void exit (int status);


在`stdlib.h`中定义了两个返回代码常量：

```
/* We define these the same for all machines.
   Changes from this to the outside world should be done in `_exit'.  */
#define	EXIT_FAILURE	1	/* Failing exit status.  */
#define	EXIT_SUCCESS	0	/* Successful exit status.  */
```

在`main`函数使用`return`语句等同于使用返回值作为参数调用`exit`函数

## abort

参考：

[abort Function](https://docs.microsoft.com/en-us/cpp/cpp/abort-function?view=vs-2019)

[abort](http://www.cplusplus.com/reference/cstdlib/abort/?kw=abort)

声明在头文件`stdlib.h`，函数`abort`表示终止`C++`程序

>void abort() noexcept;

其和`exit`函数的区别在于

* `exit`函数允许运行时终止进程（`run-time termination processing`）发生（将会调用全局对象析构函数）
* `abort`函数立即终止程序

## return

参考：[return Statement in Program Termination (C++)](https://docs.microsoft.com/en-us/cpp/cpp/return-statement-in-program-termination-cpp?view=vs-2019)

功能上和exit函数相同

## 退出注意事项

参考:

[Using exit or return](https://docs.microsoft.com/en-us/cpp/cpp/using-exit-or-return?view=vs-2019)

* 构造：全局对象（`global object`）在`main`函数执行前进行构造
* 退出：在执行`exit`或者`return`语句后，按相反顺序执行析构函数

*如果执行`abort`函数，就不会调用对象的析构函数*

### 示例代码

```
#include <iostream>
#include <cstring>
#include <array>

using std::cout;
using std::endl;

class ShowData {
public:
    // Constructor opens a file.
    ShowData(const char *szDev) {
        filename = szDev;
        cout << "construct: " << filename << endl;

        OutputDev = fopen(szDev, "w");
    }

    // Destructor closes the file.
    ~ShowData() {
        cout << "destruct: " << filename << endl;
        fclose(OutputDev);
    }

    // Disp function shows a string on the output device.
    void Disp(char *szData) {
        fputs(szData, OutputDev);
    }

private:
    FILE *OutputDev;
    const char *filename;
};

ShowData sd3 = "sd3";

ShowData sd4 = "sd4";

int main() {
    cout << "begin" << endl;

    ShowData sd1 = "sd1";

    ShowData sd2 = "sd2";

    sd1.Disp("hello to default device\n");
    sd2.Disp("hello to file hello.dat\n");
    sd3.Disp("hello to sd3");
    sd4.Disp("hello to sd4");

    cout << "end" << endl;

//    abort();
    return 0;
}
```

**注意：上面代码中全局对象的定义顺序**

结果

```
construct: sd3
construct: sd4
begin
construct: sd1
construct: sd2
end
destruct: sd2
destruct: sd1
destruct: sd4
destruct: sd3
```

### atexit

参考：

[Using atexit](https://docs.microsoft.com/en-us/cpp/cpp/using-atexit?view=vs-2019)

[atexit](http://www.cplusplus.com/reference/cstdlib/atexit/)

在调用全局对象的析构函数之前会调用[atexit](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/atexit?view=vs-2019)函数

如果对程序指定了多个`atexit`函数，则它们都将以栈的形式按相反顺序执行（即最后一次调用将第一个执行）

示例如下：

```
...
...
void f() {
    cout << "quickly, exit will happen" << endl;
}

int main() {
    cout << "begin" << endl;

    ShowData sd1 = "sd1";

    ShowData sd2 = "sd2";

    sd1.Disp("hello to default device\n");
    sd2.Disp("hello to file hello.dat\n");
    sd3.Disp("hello to sd3");
    sd4.Disp("hello to sd4");

    cout << "end" << endl;

    atexit(f);

    return 0;
}
```

结果

```
construct: sd3
construct: sd4
begin
construct: sd1
construct: sd2
end
destruct: sd2
destruct: sd1
quickly, exit will happen
destruct: sd4
destruct: sd3
```

### abort

参考：[Using abort](https://docs.microsoft.com/en-us/cpp/cpp/using-abort?view=vs-2019)

调用[abort](https://docs.microsoft.com/en-us/cpp/cpp/using-abort?view=vs-2019)函数中止程序不仅会绕过局部和全局对象的析构函数，还会绕过`atexit`函数的调用，示例如下：

```
...
    atexit(f);

    abort();

    return 0;
}
```

结果：

```
construct: sd3
construct: sd4
begin
construct: sd1
construct: sd2
end
```

只有构造函数，没有析构函数了